packages/provider/src/api/verify.ts

```diff
@@ -0,0 +1,204 @@
+        } catch (err) {
+            return next(new ProsopoApiError('CAPTCHA.PARSE_ERROR', { context: { code: 400, error: err } }))
+        }
+        try {
+            const { dappUserSignature, blockNumber, dapp } = parsed
+
+            // Verify using the dapp pair passed in the request
+            const dappPair = env.keyring.addFromAddress(dapp)
+
+            // Will throw an error if the signature is invalid
+            verifySignature(dappUserSignature, blockNumber.toString(), dappPair)
+
+            const solution = await (parsed.commitmentId
+                ? tasks.getDappUserCommitmentById(parsed.commitmentId)
+                : tasks.getDappUserCommitmentByAccount(parsed.user))
+
+            if (!solution) {
+                tasks.logger.debug('Not verified - no solution found')
+                return res.json({
+                    [ApiParams.status]: req.t('API.USER_NOT_VERIFIED'),
+                    [ApiParams.verified]: false,
+                })
+            }
+
+            if (parsed.maxVerifiedTime) {
+                const currentBlockNumber = await getCurrentBlockNumber(tasks.contract.api)
+                const blockTimeMs = getBlockTimeMs(tasks.contract.api)
+                const timeSinceCompletion = (currentBlockNumber - solution.completedAt) * blockTimeMs
+                const verificationResponse: VerificationResponse = {
+                    [ApiParams.status]: req.t('API.USER_NOT_VERIFIED'),
+                    [ApiParams.verified]: false,
+                }
+                if (timeSinceCompletion > parsed.maxVerifiedTime) {
+                    tasks.logger.debug('Not verified - time run out')
+                    return res.json(verificationResponse)
+                }
+            }
+
+            const isApproved = solution.status === CaptchaStatus.approved
+            const response: ImageVerificationResponse = {
+                [ApiParams.status]: req.t(isApproved ? 'API.USER_VERIFIED' : 'API.USER_NOT_VERIFIED'),
+                [ApiParams.verified]: isApproved,
+                [ApiParams.commitmentId]: solution.id.toString(),
+                [ApiParams.blockNumber]: solution.requestedAt,
+            }
+            return res.json(response)
+        } catch (err) {
+            return next(new ProsopoApiError('API.BAD_REQUEST', { context: { code: 400, error: err } }))
+        }
+    })
+
+    /**
+     * Verifies a user's solution as being approved or not
+     *
+     * @param {string} user - Dapp User AccountId
+     * @param {string} dapp - Dapp Contract AccountId
+     * @param {string} blockNumber - The block number at which the captcha was requested
+     * @param {string} dappUserSignature - The signature fo dapp user
+     * @param {string} commitmentId - The captcha solution to look up
+     * @param {number} maxVerifiedTime - The maximum time in milliseconds since the blockNumber
+     */
+    router.post(ApiPaths.VerifyCaptchaSolutionUser, async (req, res, next) => {
```

Comment:
Loads of shared logic here. Extract into helper functions. There should be some verify function that does the shared functionality between these endpoints

Reply:
Yeah, I think so too